src/dftbp/dftb/rangeseponscorr.F90

```diff
@@ -0,0 +1,679 @@
+    procedure :: addLrOcGradients
+
+  end type TRangeSepOnsCorrFunc
+
+
+contains
+
+
+  !> Intitialize the onsite correction with range-separated hybrid functional module
+  subroutine RangeSepOnsCorrFunc_init(this, orb, iSquare, species, onSiteElements, tSpin, rsAlg)
+
+    !> class instance
+    type(TRangeSepOnsCorrFunc), intent(out) :: this
+
+    !> Atomic orbital information
+    type(TOrbitals), intent(in) :: orb
+
+    !> Position of each atom in the rows/columns of the square matrices. Shape: (nAtom)
+    integer, dimension(:), intent(in) :: iSquare
+
+    !> species of all atoms in the system
+    integer, intent(in) :: species(:)
+
+    !> Correction to energy from on-site matrix elements
+    real(dp), intent(in), allocatable :: onSiteElements(:,:,:,:)
+
+    !> Is this spin restricted (F) or unrestricted (T)
+    logical, intent(in) :: tSpin
+
+    !> algorithm for range separation screening
+    integer, intent(in) :: rsAlg
+
+    call initAndAllocate(this, orb, iSquare, species, onSiteElements, tSpin, rsAlg)
+    call checkRequirements(this)
+
+  contains
+
+    !> initialise data structures
+    subroutine initAndAllocate(this, orb, iSquare, species, onSiteElements, tSpin, rsAlg)
+
+      !> class instance
+      class(TRangeSepOnsCorrFunc), intent(out) :: this
+
+      !> Atomic orbital information
+      type(TOrbitals), intent(in) :: orb
+
+      !> Position of each atom in the rows/columns of the square matrices. Shape: (nAtom)
+      integer, dimension(:), intent(in) :: iSquare
+
+      !> species of all atoms in the system
+      integer, intent(in) :: species(:)
+
+      !> Correction to energy from on-site matrix elements
+      real(dp), intent(in), allocatable :: onSiteElements(:,:,:,:)
+
+      !> Is this spin restricted (F) or unrestricted (T)
+      logical, intent(in) :: tSpin
+
+      !> algorithm for range separation screening
+      integer, intent(in) :: rsAlg
+
+      real(dp) :: fac
+      integer :: nAtom, iAtom, nOrb, ii, jj, iOrb, jOrb, iSh1, iSh2
+
+      nOrb = orb%nOrb
+      nAtom = size(iSquare,dim=1) - 1
+
+      this%lrOcEnergy = 0.0_dp
+      this%rsAlg = rsAlg
+      this%tSpin = tSpin
+
+      ! Set onsite constant matrices
+      allocate(this%Omat0(nOrb,nOrb))
+      allocate(this%OmatRI(nOrb,nOrb))
+      this%Omat0(:,:) = 0.0_dp
+      this%OmatRI(:,:) = 0.0_dp
+
+      do iAtom = 1, nAtom
+
+        ii = iSquare(iAtom)
+        jj = iSquare(iAtom) + orb%nOrbAtom(iAtom) - 1
+
+        do iOrb = ii, jj
+
+          ! Find l-shell for iOrb
+          iSh1 = orb%iShellOrb(iOrb-ii+1,species(iAtom))
+
+          do jOrb = ii, jj
+
+            ! Find l-shell for jOrb
+            iSh2 = orb%iShellOrb(jOrb-ii+1,species(iAtom))
+
+            ! OC contribution except diagonal elements
+            if (iOrb /= jOrb) then
+              this%Omat0(iOrb,jOrb) = onSiteElements(iSh1,iSh2,3,species(iAtom))
+            end if
+
+            ! RI loss correction for p orbitals
```

Comment:
Is there a way to generalise for other angular momentum? The original onsite code was coded for general l and worked for d and probably f. If it is not possible, there needs to be trapping for attempting to use the correction with un-supported angular momentum values.

Reply:
Yes, there is a generalisation for other angular momentum. It can be extended to d or f orbitals, but the exact equations are not derived at the moment. I will temporarily add new constraint for the d or f angular momentum.

Reply:
@bhourahine The constraint for maximum angular momentum is added, please see the commit 2d7b4eb.